### DISCUSSION TITLE
Can I use this to open audio and video inside a PDF that requires Flash?

### DISCUSSION THREAD

**Discussion by juanmiret** [ruffle-rs/ruffle#8805] [Q&A]
I have a really old PDF file that has videos and audios inside, but it requires Flash player to open them. Can I use ruffle to play them somehow?

**Comment by ousia** [comment]
@juanmiret, I don’t think this is possible, since Ruffle doesn’t understand PDF documents.

But I wonder whether the latest version of Acrobat cannot play this.

In any case, I think you may be able to extract videos and audios from the PDF file and play them as standalone media files.

**Reply by juanmiret** [reply]
Acrobat can open the PDF but cannot open the videos because they are in flash format. It asks to install flash to play them but it's not available anymore.

I tried extracting the flash content from the PDFs but it is password protected and veeery old, I don't have the pass.

I guess the only way to play them would be to find an old computer that still has flash installed?

**Reply by n0samu** [reply]
You can look into installing Clean Flash Player, which still works today: https://gitlab.com/cleanflash/installer

**Comment by ousia** [comment]
There may be another approach with [QPDF](https://github.com/qpdf/qpdf/releases/latest/).

PDF files may have two passwords:

* User password, to open the document.
* Owner password, to set restrictions on document usage.

If you may open your file without a password, it seems that your PDF document has only an owner password set.

So, you may list the attachments with:

```
qpdf --list-attachments doc.pdf
```

My sample file gives the following output:

```
video.mp4 -> 30,0
```

So I may extract the file with:

```
qpdf --show-attachment=video.mp4 doc.pdf > extracted-video.mp4
```

You need the key (the name of the attachment inside the PDF document) to redirect the output to a file.

I hope this fixes your issue.

**Reply by juanmiret** [reply]
I tried this approach. 

The --list-attachments command says that the pdf "has no embedded files".

Playing with the library I ran the --show-xref and from what I can see the file has a compressed "stream".

I don't know if that's relevant or if I can extract that.

In any case, thanks for your help! Much appreciated.

**Comment by ousia** [comment]
@juanmiret,

I‘m afraid that a stream can be pretty anything in PDF: fonts, texts, images, metadata and many other things are streams (besides embedded files).

Of course, you may try to extract the stream, but it might be encrypted too.

BTW, I have been told that recent versions of Acrobat play multimedia directly ignoring the Flash part.

This makes sense (at least, to me). Otherwise, they would be incompatible with their own spec.

If this doesn’t work, I’m afraid I don’t know how to solve your issue (which is weirdly complicated).
